File: bindings/ruby/lib/wineole/msoffice/address.rb
module WineOLE
  module MSOffice
    # Parses the addressing DSL this wrapper inherits from msoffice.rb:
    #
    #   "[Book1]Sheet1!A1:B2"   "[:new]"   ":new!"   ":first!A1"   "A1:B2"
    #
    # Touches no COM: it is a pure string parser, so it can be exercised
    # without Excel running.
    #
    # The patterns come from msoffice.rb, which encodes Excel's actual grid
    # limits (IV/65536 for Excel 11, XFD/1048576 for 12+). Both column
    # patterns were checked against every column up to their limit and are
    # exact -- and they are the harder half.
    #
    # The two row patterns needed a fix, carried here. Every nested digit
    # range started at 1 where it should start at 0: the leading digit's
    # "no leading zero" rule was carried down to positions the digits above
    # had already constrained. That dropped row 65530 outright, and 11,111
    # rows of the 1048576 grid. Both are exhaustively verified now, and
    # test_every_row_in_the_grid_parses is what keeps them that way.
    class Address
      PTN_A_IV = /[A-H]?[A-Z]|I[A-V]/i
      PTN_ABS_A_IV = /\$?#{PTN_A_IV}/
      PTN_1_65536 = /[1-9]\d{,3}|[1-5]\d{4}|6(?:[0-4]\d{3}|5(?:[0-4]\d{2}|5(?:[0-2]\d|3[0-6])))/
      PTN_ABS_1_65536 = /\$?#{PTN_1_65536}/
      PTN_ABS_A1_IV65536 = /#{PTN_ABS_A_IV}#{PTN_ABS_1_65536}/
      PTN_RANGE_LOCAL_XL11 = /(?<range>(?:#{PTN_ABS_A1_IV65536}:)?#{PTN_ABS_A1_IV65536}|#{PTN_ABS_A_IV}:#{PTN_ABS_A_IV}|#{PTN_ABS_1_65536}:#{PTN_ABS_1_65536})/
      PTN_A_XFD = /[A-W]?[A-Z]{1,2}|X(?:[A-E][A-Z]|F[A-D])/i
      PTN_ABS_A_XFD = /\$?#{PTN_A_XFD}/
      PTN_1_1048576 = /[1-9]\d{,5}|10(?:[0-3]\d{4}|4(?:[0-7]\d{3}|8(?:[0-4]\d{2}|5(?:[0-6]\d|7[0-6]))))/
      PTN_ABS_1_1048576 = /\$?#{PTN_1_1048576}/
      PTN_ABS_A1_XFD1048576 = /#{PTN_ABS_A_XFD}#{PTN_ABS_1_1048576}/
      PTN_RANGE_LOCAL_XL12 = /(?<range>(?:#{PTN_ABS_A1_XFD1048576}:)?#{PTN_ABS_A1_XFD1048576}|#{PTN_ABS_A_XFD}:#{PTN_ABS_A_XFD}|#{PTN_ABS_1_1048576}:#{PTN_ABS_1_1048576})/
      PTN_WORKBOOK = /\[(?<workbook>(?i::new)|[^\[\]]*)\]/
      PTN_WORKBOOK_WORKSHEET = /(?<worksheet_quote>'?)#{PTN_WORKBOOK}?(?<worksheet>(?i::(?:new|first|last))|(?:[^\[\]\\\:\*\'][^\[\]\\\:\*]*)?)\k<worksheet_quote>!/
      PTN_XL11 = /\A(?:#{PTN_WORKBOOK}|#{PTN_WORKBOOK_WORKSHEET}?#{PTN_RANGE_LOCAL_XL11}?)\z/
      PTN_XL12 = /\A(?:#{PTN_WORKBOOK}|#{PTN_WORKBOOK_WORKSHEET}?#{PTN_RANGE_LOCAL_XL12}?)\z/

      attr_reader :workbook, :worksheet, :range

      # Returns nil when the string is not an address at all, so a caller can
      # fall back to treating it as a raw sheet name -- which is what
      # msoffice.rb did.
      def self.parse(str, excel_version)
        pattern = excel_version.to_f >= 12 ? PTN_XL12 : PTN_XL11
        m = pattern.match(str)
        return nil if m.nil? || m.to_s.empty?

        new(workbook: m[:workbook], worksheet: m[:worksheet], range: m[:range])
      end

      def initialize(workbook:, worksheet:, range:)
        @workbook = workbook
        @worksheet = worksheet
        @range = range
      end

      # Whether this address names a range of cells. An address that stops at
      # a sheet or a book is a lookup, not an assignment target -- see the
      # spec on why `xl["Sheet1!"] = 0` filling a whole sheet is a hazard
      # rather than a convenience.
      def range?
        !@range.nil?
      end
    end
  end
end

File: bindings/python/wineole/msoffice/address.py
import re


def _group(pattern):
    """Wrap a fragment so concatenating it cannot leak an alternation.

    Ruby interpolates a Regexp into another one as `(?-mix:...)`, which
    groups it. Python's f-string interpolation does not, so
    `A|B` + `C|D` would parse as `A | BC | D`. Every fragment below is
    therefore wrapped here, once, rather than by hand at each use.
    """
    return '(?:' + pattern + ')'


# Excel 11's grid: columns A..IV, rows 1..65536.
_A_IV = _group('(?i:[A-H]?[A-Z]|I[A-V])')
_ABS_A_IV = _group(r'\$?' + _A_IV)
_1_65536 = _group(r'[1-9]\d{0,3}|[1-5]\d{4}|6(?:[0-4]\d{3}|5(?:[0-4]\d{2}|5(?:[0-2]\d|3[0-6])))')
_ABS_1_65536 = _group(r'\$?' + _1_65536)
_ABS_A1_IV65536 = _group(_ABS_A_IV + _ABS_1_65536)
_RANGE_XL11 = (
    '(?P<range>'
    + _group(_ABS_A1_IV65536 + ':') + '?' + _ABS_A1_IV65536
    + '|' + _ABS_A_IV + ':' + _ABS_A_IV
    + '|' + _ABS_1_65536 + ':' + _ABS_1_65536
    + ')'
)

# Excel 12's grid: columns A..XFD, rows 1..1048576.
_A_XFD = _group('(?i:[A-W]?[A-Z]{1,2}|X(?:[A-E][A-Z]|F[A-D]))')
_ABS_A_XFD = _group(r'\$?' + _A_XFD)
_1_1048576 = _group(
    r'[1-9]\d{0,5}|10(?:[0-3]\d{4}|4(?:[0-7]\d{3}|8(?:[0-4]\d{2}|5(?:[0-6]\d|7[0-6]))))'
)
_ABS_1_1048576 = _group(r'\$?' + _1_1048576)
_ABS_A1_XFD1048576 = _group(_ABS_A_XFD + _ABS_1_1048576)
_RANGE_XL12 = (
    '(?P<range>'
    + _group(_ABS_A1_XFD1048576 + ':') + '?' + _ABS_A1_XFD1048576
    + '|' + _ABS_A_XFD + ':' + _ABS_A_XFD
    + '|' + _ABS_1_1048576 + ':' + _ABS_1_1048576
    + ')'
)

_WORKBOOK = r'\[(?P<workbook>(?i::new)|[^\[\]]*)\]'
_WORKBOOK_WORKSHEET = (
    "(?P<worksheet_quote>'?)"
    + _group(_WORKBOOK) + '?'
    + "(?P<worksheet>(?i::(?:new|first|last))|(?:[^\\[\\]\\\\:\\*'][^\\[\\]\\\\:\\*]*)?)"
    + '(?P=worksheet_quote)!'
)


def _compile(range_pattern):
    return re.compile(r'\A' + _group(_WORKBOOK_WORKSHEET) + '?' + _group(range_pattern) + r'?\Z')


class Address:
    """Parses the addressing DSL:

        "[Book1]Sheet1!A1:B2"   "[:new]"   ":new!"   ":first!A1"   "A1:B2"

    Touches no COM: a pure string parser, so it can be exercised without
    Excel running.

    Two anchored patterns rather than Ruby's one, and that is forced:
    Ruby's `/\\A(?:WORKBOOK|WORKBOOK_WORKSHEET?RANGE?)\\z/` names the
    `workbook` group in both branches, and Python's `re` refuses a group
    name that appears twice in one pattern. The book-only pattern is
    therefore tried first and the full one second -- the same order Ruby's
    alternation tries them in, so the acceptance set and the captures are
    identical. The address test file proves it with every form of the Ruby
    test file plus the grid boundaries.
    """

    PTN_BOOK_ONLY = re.compile(r'\A' + _WORKBOOK + r'\Z')
    PTN_XL11 = _compile(_RANGE_XL11)
    PTN_XL12 = _compile(_RANGE_XL12)

    def __init__(self, workbook, worksheet, range):
        self.workbook = workbook
        self.worksheet = worksheet
        self.range = range

    @staticmethod
    def parse(text, excel_version):
        """Returns None when the string is not an address at all, so a
        caller can fall back to treating it as a raw sheet name -- which is
        exactly what Excel._lookup does."""
        if not isinstance(text, str):
            return None

        m = Address.PTN_BOOK_ONLY.match(text)
        if m is not None:
            return Address(m.group('workbook'), None, None)

        pattern = Address.PTN_XL12 if Address._is_xl12(excel_version) else Address.PTN_XL11
        m = pattern.match(text)
        # An empty match means the pattern's every part was optional and
        # every one of them was absent -- "" is not an address.
        if m is None or m.group(0) == '':
            return None
        return Address(m.group('workbook'), m.group('worksheet'), m.group('range'))

    @staticmethod
    def _is_xl12(excel_version):
        # Ruby reads the version with String#to_f, which answers 0.0 for
        # anything unparseable rather than raising. Same leniency here: a
        # version this code cannot read is treated as the smaller grid,
        # which refuses more addresses rather than accepting ones the host
        # cannot hold.
        try:
            return float(excel_version) >= 12
        except (TypeError, ValueError):
            return False

    @property
    def has_range(self):
        """Whether this address names a range of cells. An address that
        stops at a sheet or a book is a lookup, not an assignment target --
        `xl["Sheet1!"] = 0` filling a whole sheet is a hazard rather than a
        convenience."""
        return self.range is not None
